osquery/sql/sqlite_collation.cpp

```diff
@@ -0,0 +1,257 @@
+  default:
+    return false;
+  }
+}
+
+/**
+ * @brief Set the delimiter precedence if they should compare against each
+ * other. Delimiters: `~-^.:`
+ */
+static int delimiterPrecedence(int d) {
+  switch (d) {
+  case 126:
+    return 1;
+  case 45:
+    return 2;
+  case 94:
+    return 3;
+  case 46:
+    return 4;
+  case 58:
+    return 5;
+  default:
+    return 0;
+  }
+}
+
+/**
+ * @brief Return remainder sort order, or return the version length difference.
+ */
+static int compareRemainder(int lLen,
+                            const char* lVer,
+                            int rLen,
+                            const char* rVer,
+                            int pos,
+                            int diff,
+                            const bool comp_remaining,
+                            const bool remainder_precedence) {
+  // This supports linux package versioning sort order when a tilde should be
+  // less than, a caret should be greater than, and a hyphen should be equal.
+  //
+  // When remainder_precedence = true, this will return the segment value diff
+  // if the last compared character was numeric, otherwise it falls through to
+  // return length diff.
+  //
+  // When remainder_precedence = false, this will return the segment value diff
+  // if there is any before falling through to return length diff.
+  if (comp_remaining) {
+    if (lLen == pos) {
+      switch (int(rVer[pos])) {
+      case 126:
+        return 1;
+      case 45:
+        return 0;
+      case 94:
+        return -1;
+      default:
+        if (diff != 0 && (!remainder_precedence || !isdigit(lVer[pos - 1]))) {
+          return diff;
+        }
+      }
+    }
+
+    if (rLen == pos) {
+      switch (int(lVer[pos])) {
+      case 126:
+        return -1;
+      case 45:
+        return 0;
+      case 94:
+        return 1;
+      default:
+        if (diff != 0 && (!remainder_precedence || !isdigit(rVer[pos - 1]))) {
+          return diff;
+        }
+      }
+    }
+  }
+
+  return lLen - rLen;
+}
+
+/**
+ * @brief Compares two versions strings against each other.
+ * Return 0 if the versions should evaluate as equal.
+ * Return negative int if the left string is less than the right.
+ * Return positive int if the left string is greater than the right.
+ */
+static int versionCompare(int lLen,
+                          const void* lVersion,
+                          int rLen,
+                          const void* rVersion,
+                          const bool epoch = false,
+                          const bool delim_precedence = false,
+                          const bool comp_remaining = false,
+                          const bool remainder_precedence = false) {
+  const char* lVer = static_cast<const char*>(lVersion);
+  const char* rVer = static_cast<const char*>(rVersion);
+
+  // Early return if versions are equal.
+  if (lVer == rVer) {
+    return 0;
+  }
+
+  // Check for and return difference in epoch position.
+  if (epoch) {
```

Comment:
I _suspect_ the epoch check can be done inside the `if (lDelim && !rDelim)` blocks. But maybe this is okay.

Reply:
Yeah that could work as well. I wasn't entirely sure if my epoch check makes the most sense where it is, so I can play around with it.